File: python/04_queue/circular_queue.py
"""
基于数组的循环队列
"""
from itertools import chain


class CircularQueue:
    def __init__(self, capacity: int):
        self.items = []
        self._capacity = capacity
        self._head = 0
        self._tail = 0

    def is_full(self):
        return (self._tail + 1) % self._capacity == self._head

    def is_empty(self):
        return self._tail == self._head

    def enqueue(self, item: str) -> bool:
        if self.is_full():
            return False
        self.items.append(item)
        self._tail = (self._tail + 1) % self._capacity
        return True

    def dequeue(self) -> str:
        if not self.is_empty():
            item = self.items[self._head]
            self._head = (self._head + 1) % self._capacity
            return item

    def __repr__(self):
        if self._tail >= self._head:
            return " ".join(self.items[self._head:self._tail])
        else:
            return " ".join(chain(self.items[self._head:], self.items[:self._tail]))


if __name__ == '__main__':

    queue = CircularQueue(5)

    for i in range(5):
        queue.enqueue(str(i))
    print(queue)

    for _ in range(3):
        print(queue.dequeue())

    queue.enqueue('a')
    queue.enqueue('b')
    for _ in range(3):
        queue.dequeue()
    print(queue)

File: go/04_queue/circular_queue.go
package _4_queue

import (
	"fmt"
	"strings"
)

type CircularQueue struct {
	data     []interface{}
	head     int
	tail     int
	capacity int
}

func NewCircularQueue(n int) *CircularQueue {
	return &CircularQueue{make([]interface{}, n), 0, 0, n}
}

// 当队列满时， tail 指向的位置实际上是没有存储数据的。所以，循环队列会浪费一个数组的存储空间。

func (c *CircularQueue) IsFull() bool {
	if c.head == (c.tail+1)%c.capacity {
		return true
	}
	return false
}

func (c *CircularQueue) IsEmpty() bool {
	if c.head == c.tail {
		return true
	}
	return false

}

func (c *CircularQueue) EnQueue(v interface{}) bool {
	if c.IsFull() {
		return false
	}
	c.data[c.tail] = v
	c.tail = (c.tail + 1) % c.capacity
	return true
}

func (c *CircularQueue) DeQueue() interface{} {
	if c.IsEmpty() {
		return nil
	}
	v := c.data[c.head]
	c.head = (c.head + 1) % c.capacity
	return v
}

func (c *CircularQueue) Print() string {

	result := ""

	i := c.head
	for {
		result += fmt.Sprintf("%v->", c.data[i])
		i = (i + 1) % c.capacity
		if i == c.tail {
			break
		}

	}

	result = strings.TrimSuffix(result, "->")
	return result

}
